File: src/week6/P005.java
package week6;

import java.io.*;
import java.util.StringTokenizer;

public class P005 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        // K가 0이거나 N인 경우 이항 계수는 항상 1
        if (K == 0 || K == N) {
            System.out.println(1);
            return;
        }

        int[][] pascal = new int[N + 1][];

        if (N > 1) {
            pascal[1] = new int[]{1, 1};
        }

        // DP 수행
        for (int depth = 2; depth < N; depth++) {
            // 해당 depth(행)의 크기는 depth + 1이어야 함 (예: 2번째 행은 1, 2, 1로 3개)
            pascal[depth] = new int[depth + 1];

            pascal[depth][0] = 1; // 첫 번째 원소

            // 중간 원소들 계산
            for (int idx = 1; idx < depth; idx++) {
                pascal[depth][idx] = (pascal[depth - 1][idx - 1] + pascal[depth - 1][idx]) % 10007;
            }

            pascal[depth][depth] = 1; // 마지막 원소
        }

        int result = (pascal[N - 1][K - 1] + pascal[N - 1][K]) % 10007;
        System.out.println(result);
    }
}

File: src/week6/P005.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split(' ');

const N = parseInt(input[0]);
const K = parseInt(input[1]);

if (K === 0 || K === N) {
    console.log(1);
} else {
    // 파스칼 삼각형 배열 초기화
    let pascal = [];
    
    // n=1 일 때 [1, 1]
    pascal[1] = [1, 1];

    for (let depth = 2; depth < N; depth++) {
        // 자바스크립트는 배열 크기를 미리 지정하지 않아도 되지만, 
        // 명시적으로 depth + 1 크기의 배열을 생성하는 것이 로직상 안전함
        pascal[depth] = new Array(depth + 1);

        pascal[depth][0] = 1; // 첫 원소

        for (let idx = 1; idx < depth; idx++) {
            // 모듈러 연산 적용
            pascal[depth][idx] = (pascal[depth - 1][idx - 1] + pascal[depth - 1][idx]) % 10007;
        }

        pascal[depth][depth] = 1; // 마지막 원소
    }

    // 결과 출력
    console.log((pascal[N - 1][K - 1] + pascal[N - 1][K]) % 10007);
}